### ISSUE TITLE
Allow removing all data on uninstall

### ISSUE THREAD

**Issue by swissspidy** [ampproject/amp-wp#3210]
As reported in the forums: https://wordpress.org/support/topic/how-to-delete-plugin-including-data-left-behind/

When uninstalling the plugin, all the AMP options and custom post types are still in the database.

Suggestion:

1. Add checkbox to AMP settings screen to enable data removal upon uninstall
2. Use `register_uninstall_hook` or `uninstall.php` to remove all custom post types, taxonomies, and options upon uninstall.

**Comment by westonruter** [comment]
There's also the entry on `wp_options`, and the postmeta for controlling whether AMP is enabled for a given post.

Is the checkbox necessary? Doesn't uninstallation entail data removal? I'm not up on the best practices for plugin uninstallation.

**Comment by swissspidy** [comment]
A checkbox is common in many plugins as it prevents accidental data removal, for example when one wants to update a plugin via uploading a ZIP file. Because https://core.trac.wordpress.org/ticket/9757 is not solved yet, it would erase all data when deleting the AMP plugin and uploading the ZIP file of an RC. Which would be a bit annoying.

**Comment by cryptochrome** [comment]
I can't comprehend why this wasn't built into the plugin from the get go. When you guys implement this, please also consider taking care of redirects (301) required after removing all AMP content. You don't want to leave your users with potentially thousands of 404's.

**Comment by westonruter** [comment]
@cryptochrome What 404s are you referring to? Are you referring to `/amp/` endpoint URLs? There would be no way for the AMP plugin to redirect those URLs if it is deactivated. WordPress has no persistent URL redirect system that it could write to upon deactivation. 

The issue of `/amp/` URLs 404ing will be mitigated by away from URL endpoints in favor of the `?amp=1` query var per https://github.com/ampproject/amp-wp/issues/2204.

**Comment by westonruter** [comment]
@cryptochrome To redirect `/amp/` URLs to non-AMP URLs after plugin deactivation, you can use this plugin: https://gist.github.com/westonruter/b7eb9cc7648b9a8b94b0015c79c8702e

**Comment by westonruter** [comment]
Something that comes to mind with this issue is that if someone wants to try a pre-release build of the AMP plugin, currently the way to do that is to download a pre-release ZIP, uninstall the AMP plugin, and install the pre-release AMP plugin to then activate. However, if the uninstall process deletes data, then this would not work anymore.

**Comment by swissspidy** [comment]
@westonruter That‘s why I mention the checkbox. Uninstall would only run if the user enables it.

**Comment by swissspidy** [comment]
> Something that comes to mind with this issue is that if someone wants to try a pre-release build of the AMP plugin, currently the way to do that is to download a pre-release ZIP, uninstall the AMP plugin, and install the pre-release AMP plugin to then activate.

This is fixed in WordPress 5.5. Updating via upload is now possible.

Time to unblock?

**Comment by westonruter** [comment]
Yes, let's do it in v2.1.

**Comment by kmyram** [comment]
This needs an Implementation Brief or similar to that effect

**Comment by westonruter** [comment]
Implementation brief:

Add an `uninstall.php` which does what @swissspidy outlined in the [support topic](https://wordpress.org/support/topic/how-to-delete-plugin-including-data-left-behind/#post-11913320).

1. Remove AMP settings data: `wp option delete amp-options`
2. Remove all `amp_validated_url` posts: `wp post delete $(wp post list --post_type='amp_validated_url' --format=ids)`
3. Remove all `amp_validation_error` terms: `wp term list amp_validation_error --field=term_id | xargs wp term delete amp_validation_error`

We could also delete all transients associated with the plugin, when an external object cache is not being used.

See also [Uninstall Methods](https://developer.wordpress.org/plugins/plugin-basics/uninstall-methods/) in the Plugin Handbook.

**Comment by westonruter** [comment]
Something that comes to mind with the switch to using custom tables for validation data storage (#5553) is this would greatly facilitate deletion of the validation data upon uninstallation. No longer would we need to call `wp_delete_post()` and `wp_delete_term()` repeatedly. We could instead just `DROP TABLE` for each, which would be much faster.

**Comment by amanijoseph87** [comment]
Is this implemented yet..!

**Comment by westonruter** [comment]
No, but it's planned for v2.2—the next release after v2.1 which we are publishing this week. Sorry for the delay.

**Comment by amanijoseph87** [comment]
Can you help me how to remove this data from my site, i have try to delete some table on my database without success, is there anyway we can use while we wait..

**Comment by westonruter** [comment]
@amanijoseph87 This plugin does not create any database tables, at least not yet. Are you sure you're using this AMP plugin?

**Comment by amanijoseph87** [comment]
Am using this plugin, even on board wizard display blank.

**Comment by westonruter** [comment]
@amanijoseph87 OK, please open a support topic on the forum. Our team will be able to better help you there: https://wordpress.org/support/plugin/amp/#new-topic-0

**Comment by amanijoseph87** [comment]
And settings page display blank too nothing seems to work

**Comment by amanijoseph87** [comment]
Please check - https://wordpress.org/support/?post_type=topic&p=14395574

**Comment by amanijoseph87** [comment]
Is there a way to reset this plugin as delete all data related to this plugin

**Comment by westonruter** [comment]
That link takes me to a 404 page. Please open a topic here: https://wordpress.org/support/plugin/amp/#new-topic-0

We'll see it once you've posted it. 

Otherwise, you can follow the manual steps outlined in https://github.com/ampproject/amp-wp/issues/3210#issuecomment-679325326 until we add an `uninstall.php`.

**Comment by amanijoseph87** [comment]
( https://wordpress.org/support/?post_type=topic&p=14395574 ) Is held for moderation i don't know way

**Comment by westonruter** [comment]
OK, we'll follow up with you there then.

**Comment by amanijoseph87** [comment]
Please do now i know you guys a busy and if its not now i can not reach you in other time

**Comment by westonruter** [comment]
We'll continue this on the support forum.

**Comment by westonruter** [comment]
⚠️⚠️⚠️  This will need careful QA since it involves deletion of data! ⚠️⚠️⚠️

See how I initially tested during review in https://github.com/ampproject/amp-wp/pull/6422#pullrequestreview-714176855.

**Comment by westonruter** [comment]
Follow-up discussion on whether there should ba a mechanism for explicitly opting-in to deleting data: https://github.com/google/web-stories-wp/issues/8453

We could add a new option on the settings screen which is disabled by default to ”Delete all plugin data upon uninstallation“. This could be disabled by default if we want to be extra cautious. If a user uninstalls the plugin and finds the data wasn't removed, they could re-install and then enable that option and re-uninstall.

Nevertheless, the data being removed (e.g. validated URLs) is much less important than the data being stored by Web Stories.

**Comment by bartoszgadomski** [comment]
**QA passed:**

Only AMP-related data has been removed from database; no AMP-related data left in database after removal. Attaching "before.sql" and "after.sql" DB dumps that I used to compare: [db-dumps.zip](https://github.com/ampproject/amp-wp/files/7677128/db-dumps.zip)

![Screenshot 2021-12-8 at 16 30 16](https://user-images.githubusercontent.com/4476915/145237227-c73cdb74-8254-40cd-b782-1693fce8f076.png)
![Screenshot 2021-12-8 at 16 30 05](https://user-images.githubusercontent.com/4476915/145237230-f0c1b129-e60e-45cb-a4c7-78fa46f588aa.png)
![Screenshot 2021-12-8 at 16 29 41](https://user-images.githubusercontent.com/4476915/145237236-1a4da11d-1ccc-45aa-b48b-31cc5fb50218.png)
![Screenshot 2021-12-8 at 16 32 24](https://user-images.githubusercontent.com/4476915/145237239-5688b263-45dc-48b7-9f21-a1ebcc50565b.png)
![Screenshot 2021-12-8 at 16 32 55](https://user-images.githubusercontent.com/4476915/145237245-09102b4a-f1b0-4f7b-9870-e91e076e77b1.png)
